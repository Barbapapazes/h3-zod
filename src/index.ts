import type { CompatibilityEvent } from 'h3'
import { createError, useBody, useQuery } from 'h3'
import { z } from 'zod'

// copy of the private Zod utility type of ZodObject
type UnknownKeysParam = 'passthrough' | 'strict' | 'strip'

type TQuery<U extends UnknownKeysParam = any> =
  | z.ZodObject<any, U>
  | z.ZodUnion<[TQuery<U>, ...TQuery<U>[]]>
  | z.ZodIntersection<TQuery<U>, TQuery<U>>
  | z.ZodDiscriminatedUnion<string, z.Primitive, z.ZodObject<any, U>>

export function useValidatedQuery<T extends TQuery>(
  event: CompatibilityEvent,
  schema: T,
) {
  const query = useQuery(event)
  const parsed = schema.safeParse(query)

  if (!parsed.success) {
    throw createError({
      statusCode: 400,
      statusMessage: JSON.stringify(parsed.error.errors),
    })
  }

  return parsed.data as z.infer<T>
}

export async function useValidatedBody<T extends TQuery>(
  event: CompatibilityEvent,
  schema: T,
) {
  const body = await useBody(event)
  const parsed = schema.safeParse(body)

  if (!parsed.success) {
    throw createError({
      statusCode: 400,
      statusMessage: JSON.stringify(parsed.error.errors),
    })
  }

  return parsed.data as z.infer<T>
}

export {
  z,
}
