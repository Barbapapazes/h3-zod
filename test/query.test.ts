import { beforeEach, describe, expect, it } from 'vitest'
import type { SuperTest, Test } from 'supertest'
import supertest from 'supertest'
import type { App, CompatibilityEvent } from 'h3'
import { createApp } from 'h3'
import { useValidatedQuery, z } from '../src'

describe('useValidatedQuery', () => {
  let app: App
  let request: SuperTest<Test>

  beforeEach(() => {
    app = createApp({ debug: false })
    request = supertest(app)
  })

  const querySchema = z.object({
    required: z.string(),
  })

  it('returns 200 OK if query matches validation schema', async () => {
    app.use('/validate', (req: CompatibilityEvent) => useValidatedQuery(req, querySchema))

    const res = await request.get('/validate?required')

    expect(res.status).toEqual(200)
    expect(res.body).toMatchSnapshot()
  })

  it('throws 400 Bad Request if query does not match validation schema', async () => {
    app.use('/validate', (req: CompatibilityEvent) => useValidatedQuery(req, querySchema))

    const res = await request.get('/validate')

    expect(res.status).toEqual(400)
    expect(res.body).toMatchSnapshot()
  })
})
