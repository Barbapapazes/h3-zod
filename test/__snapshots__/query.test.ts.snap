// Vitest Snapshot v1

exports[`useValidatedQuery > doesn't throw 400 Bad Request if query does not match validation schema 1`] = `
{
  "error": {
    "issues": [
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "required",
        ],
        "received": "undefined",
      },
    ],
    "name": "ZodError",
  },
  "success": false,
}
`;

exports[`useValidatedQuery > returns 200 OK if query matches validation schema 1`] = `
{
  "required": "",
}
`;

exports[`useValidatedQuery > throws 400 Bad Request if query does not match validation schema 1`] = `
{
  "data": "{\\"issues\\":[{\\"code\\":\\"invalid_type\\",\\"expected\\":\\"string\\",\\"received\\":\\"undefined\\",\\"path\\":[\\"required\\"],\\"message\\":\\"Required\\"}],\\"name\\":\\"ZodError\\"}",
  "stack": [],
  "statusCode": 400,
  "statusMessage": "Bad Request",
}
`;
